libp2p/switch.nim

```diff
@@ -425,17 +428,18 @@ proc accept(s: Switch, transport: Transport) {.async.} = # noraises
     try:
       debug "About to accept incoming connection"
       conn = await transport.accept()
-      if not isNil(conn):
-        debug "Accepted an incoming connection", conn
-        asyncSpawn s.upgradeIncoming(conn) # perform upgrade on incoming connection
-      else:
+      if isNil(conn):
         # A nil connection means that we might have hit a
         # file-handle limit (or another non-fatal error),
         # we can get one on the next try, but we should
         # be careful to not end up in a thigh loop that
         # will starve the main event loop, thus we sleep
         # here before retrying.
         await sleepAsync(100.millis) # TODO: should be configurable?
+        continue
+
+      debug "Accepted an incoming connection", conn
+      await s.upgradeIncoming(conn) or conn.upgraded # wait for connection to by upgraded
```

Comment:
how can the connection be upgraded without upgradeIncoming completing? why are both futures needed here?

Reply:
`upgradeIncoming` doesn't complete until the connection exists because it awaits the multistream handler which dispatches calls to user protocol handlers here - https://github.com/status-im/nim-libp2p/blob/c788a6a3c0ca614be4c67cc566aee6877da0250b/libp2p/switch.nim#L356, that's why I'm adding another event to the connection because the upgrade completes in the `muxerHandler` here https://github.com/status-im/nim-libp2p/blob/c788a6a3c0ca614be4c67cc566aee6877da0250b/libp2p/switch.nim#L624 (when added to the `ConnectionManager`).

Reply:
seems easier to fail `conn.upgraded` with an exception then, `iff` upgrade fails in such a way that `upgradeIncoming` ends before `upgraded`

Reply:
basically, we should _never_ leave futures around that don't finish -  that's a prime source of async leaks

Reply:
Sure, cleaning up/canceling the `.upgraded` future on a failed `upgradeIncoming` sounds like a reasonable thing to do.

Reply:
I just though that this `or` construct isn't necessary at all as I can just:

```nim
asyncSpawn s.upgradeIncoming(conn)
await conn.upgraded
``` 

and simply complete/fail the `conn.upgraded` if upgrade failed.